ipcache: Add cilium monitor events and expose it via API

gandro opened the pull request:
This PR contains two new additions:

First, it adds a new API endpoint, `/ip`, which allows external clients to dump the current content of the user-level IP cache. This allows them to efficiently map IP addresses to identities and (if available) to Kubernetes pod names.

Second, it extends cilium monitor with two new `AgentNotify` messages, `AgentNotifyIPCacheUpserted` and `AgentNotifyIPCacheDeleted` which inform consumers of the cilium monitor socket about updates to the IP cache.

Review per commit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/9268)
<!-- Reviewable:end -->

joestringer commented on daemon/ipcache.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+type getIPCache struct {
+	daemon *Daemon
```
Is this field used anywhere?

joestringer commented on daemon/ipcache.go:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	return NewGetIPCacheOK().WithPayload(&models.IPCache{
+		Cache: listener.entries,
```
Is there a maximum payload length here? If we have a really big ipcache how will this behave?

joestringer replied:
(Are there similar examples of large dumps we may do in the API already?)

gandro replied:
That's a very fair point. There is no maximum length, but there are similar examples: I modeled the API to roughly mirror `/fqdn/cache`, which also does not have a limit. But I'm not sure how the IP cache compares to the FQDN cache in terms of size and update rates. While it depends on the workload, I'm guessing the IP cache will have at least as many entries and updates as the FQDN cache?

In terms of impact, dumping the cache via API acquires a read-lock, which naturally will block any IP cache updates. Not sure what we can do here to avoid . My use case for the API is to obtain a initial snapshot of the IP cache. 

Thinking about it, splitting the API into the typical `list keys` (which should be cheaper) and `fetch details for key` endpoints will allow updates to be interleaved. This will of course (by design) not give me complete snapshot of the cache, but I should be fine since I can listen to monitor events and deal with invalidated entries accordingly.

gandro replied:
After looking at other API endpoints, I'm less sure about splitting the API.

All other collection paths (`/endpoint`, `/identity`, `/service`, the aforementioned `/fqdn/cache`) dump fully populated lists upon request. They arguably all contain less entries than the IP cache, but the response body for some of them (e.g. `/endpoint`) will typically likely be larger than the dumped IP cache, as they contain more fields.

joestringer replied:
FQDN cache is a good question; I forget whether these are generated based upon the number of FQDN rules in the policy, or based upon the number of unique FQDNs seen by the DNS proxy. I suspect it's the former, which would limit it considerably.

Looking at endpoint, I don't think we're expecting more than perhaps 100 endpoints on a given node (which even that is quite high compared to typical deployments today). Looking at the ipcache, the default maximum is 512,000. While the endpoint is a more complicated structure, the order of magnitude is very different.

Identities is probably closer to the scale of ipcache, while I don't think that there is a BPF map that tells us the limit, we've investigated in the order of tens of thousands.

One thing you might consider doing to test this out and just generally make sure the API behaves correctly is to inject high numbers of IP-ID entries into the kvstore. I did something like that here: https://github.com/cilium/cilium/issues/8063#issuecomment-493236355

The read-lock question is mainly interesting if it's going to happen particularly often. Do you have a sense for your use case how frequently you want to hit that API? There are mechanisms to make that behaviour better if necessary, but it's quite a bit more work.

tgraf replied:
Given that we only hold the lock to dump into memory, I think it is fine.

gandro replied:
> One thing you might consider doing to test this out and just generally make sure the API behaves correctly is to inject high numbers of IP-ID entries into the kvstore. I did something like that here: #8063 (comment)

Oh, that's a great idea. I'll give it a shot.

> Do you have a sense for your use case how frequently you want to hit that API? There are mechanisms to make that behaviour better if necessary, but it's quite a bit more work.

My goal is to maintain a history of the ipcache in downstream consumers of the cilium monitor socket. I think I can reduce the API load a lot if I rely on monitor socket notifications. If I fetch the ipcache snapshot after connecting to the monitor socket I should be able maintain my copy of the ipcache via notifications only and do not need to consult the API at all after that (until the monitor socket gets disconnected for whatever reason of course). 

(This assumes I can detect update notifications older than my initial snapshot as such, which I think I can since the upserts contain the old values they replace). If not, we would have to add a node-local revision field to the ipcache)

gandro replied:
@joestringer I ran a rough benchmark yesterday to get an idea how expensive the RLock is, the numbers are from a VM running on my dev laptop, so take them with a grain of salt as there was likely some background noise.

I ran the experiment with 25, 50 and 75k entries in the ipcache. I was a bit surprised to see that it is indeed not negligible. With 75k ipcache entries, the lock is held on average 92 milliseconds. 

Percentiles (all values in milliseconds):

| IPCache Entries:      |         25k |      50k |      75k |
|:------|------------:|---------:|---------:|
| mean  |    25.2579  |  56.7184 |  92.4482 |
| std   |     7.55403 |  10.8656 |  22.2895 |
| 50%   |    21.8291  |  51.7066 |  86.0281 |
| 75%   |    28.3468  |  64.2931 | 103.095  |
| 99.9% |    66.3948  | 143.382  | 227.67   |
| 100%  |   422.616   | 175.278  | 641.549  |

Boxplot (values larger than 99.8%-percentile removed for readability):
![image](https://user-images.githubusercontent.com/50564/65955968-97adb500-e449-11e9-96df-af2641913d1b.png)

In conclusion, it can block ipcache updates for hunders of millisconds when using the API. In really bad cases likely more a second if we have >100k entries. As discussed, we will have to be careful with calling the new `/ip/` API endpoint to not add unnecessarily block updates.

joestringer commented on daemon/daemon.go:
```diff
@@ -521,6 +521,8 @@ func NewDaemon(dp datapath.Datapath, iptablesManager rulesManager) (*Daemon, *en
 	}
 
+	// This IPCache event listener forwards events to cilium monitor
+	ipcache.IPIdentityCache.AddListener(&d)
```
Shouldn't this occur in `initMaps()` where the other IPIdentityCache listener is added to the `ipcache.IPIdentityCache`?

gandro replied:
Ah, thanks for the pointer! I did not see that, will change.

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -800,6 +800,27 @@ paths:
           description: Invalid request (error parsing parameters)
           schema:
             "$ref": "#/definitions/Error"
+  "/ip/cache/":
```
Although we internally call this the "IP cache", do we want to also name it this way in the API? The "cache" part is a bit questionable for some parts of its information, for example CIDR identities, where this is actually the source of truth for this agent's store of this information.

More generally this is "information about an IP" where the IP is the key, then the value includes a bunch of information - identity, encryption key id, k8s pods/namespaces, etc.

Should it just be `/ip/`?

joestringer commented on daemon/ipcache.go:
```diff
@@ -0,0 +1,147 @@
+// Copyright 2016-2019 Authors of Cilium
```
Is it really copyright 2016-2019?

joestringer commented on daemon/ipcache.go:
```diff
@@ -0,0 +1,147 @@
+// Copyright 2016-2019 Authors of Cilium
+//
```
Please add this file to CODEOWNERS so `ipcache` group can review.

joestringer reviewed (changes requested):
The implementation looks pretty straightforward, thanks!

I have a few minor comments below.

joestringer commented on api/v1/openapi.yaml:
```diff
@@ -2159,3 +2180,43 @@ definitions:
+      k8sMetadata:
+        "$ref": "#/definitions/IPCacheEntryK8sMetadata"
+  IPCacheEntryK8sMetadata:
+    description: Kubernetes Metadata of an IP cache entry
```
Does this need to be k8s-specific? Wondering if it could be a little more generic. Something like "source: k8s, name: xxx, namespace: yyy"

joestringer commented on daemon/ipcache.go:
```diff
@@ -0,0 +1,147 @@
+)
+
+// OnIPIdentityCacheChange listens to ipcache updates and forwards them to the monitor
+func (d *Daemon) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr net.IPNet,
```
How tightly do we want to bind this to the actual change in datapath state?

Currently this is implemented as a separate notification that is triggered from the ipcache, which may occur before or after the actual change to BPF, with no timestamp reference as to whether the change in datapath behaviour just occurred or is about to occur. I *suspect* that it will happen just after the BPF is plumbed because the listeners are a slice and this likely implies ordering, but I don't know if this is guaranteed.

It could be more closely tied by having some integration into `pkg/datapath/ipcache`, but then I don't know in practical terms how much more accurate we will get and whether it's worth the code complexity. At some point, the syscall to modify the datapath state will occur, and the notification needs to be triggered before/after. But figured I would ask just to see if you thought about that.

gandro replied:
Good point, thanks for bringing this up! I did not consider that. The current design was mostly chosen to avoid much code coupling.

I guess without actually sending some form of ipcache update notification though the perf ring buffer there cannot be any strict order guarantees for cilium monitor clients between ipcache notifications and trace notifications (which is what we ideally want as a cilium monitor client). This in turns would mean we have to send the notification from a BPF program? That sounds like a lot of complexity.

For now, I will change the PR to emit the monitor notification in `pkg/datapath/ipcache`, as it seems relatively straight-forward to do.

gandro replied:
(As a side note, thanks as always for the amazing code review - much appreciated!)

joestringer replied:
Agreed that you would need a single channel, but I don't know that there's a really good way to emit an event into the perf event channel when the element is added based on the syscall; maybe something could be done with tracepoints but for now this is more than necessary.

joestringer reviewed (approved):
I think it's worthwhile following up on the behaviour question re: how often the API should be used and making sure that large IPcache maps are dumped properly, but other than that the PR LGTM.

tgraf commented on api/v1/openapi.yaml:
```diff
@@ -2155,3 +2179,38 @@ definitions:
         description: The endpoint that made this lookup, or 0 for the agent itself.
         type: integer
+  IPListEntry:
+    description: IP entry with metadata"
```
The `"` looks wrong

tgraf commented on daemon/daemon_main.go:
```diff
@@ -1419,5 +1419,8 @@ func (d *Daemon) instantiateAPI() *restapi.CiliumAPI {
 	api.PolicyGetFqdnCacheIDHandler = NewGetFqdnCacheIDHandler(d)
 	api.PolicyGetFqdnNamesHandler = NewGetFqdnNamesHandler(d)
 
+	// /ip/cache/
```
The API documents `/ip`, the code says `/ip/cache`. It's fixed up in the next commit though so I think it's just a small bug in how you squashed things together.

tgraf commented on daemon/ipcache.go:
```diff
@@ -0,0 +1,106 @@
+		return NewGetIPCacheNotFound()
+	}
+
+	return NewGetIPCacheOK().WithPayload(&models.IPCache{
```
The code here doesn't seem to match up with the API defined. It seems to be fixed in the next commit though. Can you fix it up so we don't break bisecting?

gandro replied:
Good catch, that was not my intention! Apparently I did a sloppy job of rewriting the git history :disappointed_relieved:

tgraf commented on pkg/datapath/ipcache/listener.go:
```diff
@@ -57,18 +63,58 @@ type BPFListener struct {
+		oldIdentityPtr           *uint32
+	)
+
+	if l.monitorNotify == nil {
```
I think this is fine for now but we may need to add a filter flag to the Monitor API to have the monitor listener subscribe to these events. We may not want that every `cilium monitor` run sees all ipcache updates in the entire cluster.

tgraf reviewed (changes requested):
Looks very good, great work. Some minor things.

gandro committed:
api: Add /ip endpoint for listing IP cache entries

This adds a new API endpoint named /ip for fetching IP cache entries
from the agent. It has a single query parameter called 'cidr' which
can be used to filter IP cache entries based on a CIDR range.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

gandro committed:
monitor: Add AgentNotify messages for IPCache updates

This adds a new ipcache listener in the daemon to forward any ipcache
updates to the monitor agent. This allows clients of the cilium monitor
socket to react to changes to the local ipcache.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>
